### PR TITLE
Allow _embed to include all embedded objects 

### PR DESCRIPTION
**Pull request by spacedmonkey**
Follow on from https://github.com/WordPress/wordpress-develop/pull/672

Trac ticket: https://core.trac.wordpress.org/ticket/43439

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/rest-api/rest-server.php [hunk 1]
```diff
@@ -931,6 +931,37 @@ public function test_link_embedding_without_links() {
 		$this->assertSame( 'data', $result['untouched'] );
 	}
 
+	/**
+	 * Ensure embedding is with links in the data.
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/rest-api/rest-server.php:935]:
```suggestion
	 * Ensure embedding is with links in the data.
	 *
	 * @ticket 43439
```

Ticket number is missing.

**Comment by moshiajonta** [comment]
<a href='https://capmistoi-6381.weebly.com'>.</a>
<a href='https://capmistoi-6382.weebly.com'>.</a>
<a href='https://capmistoi-6383.weebly.com'>.</a>
<a href='https://capmistoi-6384.weebly.com'>.</a>
<a href='https://capmistoi-6385.weebly.com'>.</a>
<a href='https://capmistoi-6386.weebly.com'>.</a>
<a href='https://capmistoi-6387.weebly.com'>.</a>
<a href='https://capmistoi-6388.weebly.com'>.</a>
<a href='https://capmistoi-6389.weebly.com'>.</a>
<a href='https://capmistoi-6390.weebly.com'>.</a>
<a href='https://capmistoi-6391.weebly.com'>.</a>
<a href='https://capmistoi-6392.weebly.com'>.</a>
<a href='https://capmistoi-6393.weebly.com'>.</a>
<a href='https://capmistoi-6394.weebly.com'>.</a>
<a href='https://capmistoi-6395.weebly.com'>.</a>
<a href='https://capmistoi-6396.weebly.com'>.</a>
<a href='https://capmistoi-6397.weebly.com'>.</a>
<a href='https://capmistoi-6398.weebly.com'>.</a>
<a href='https://capmistoi-6399.weebly.com'>.</a>
<a href='https://capmistoi-6400.weebly.com'>.</a>

File: src/wp-includes/rest-api/class-wp-rest-server.php [hunk 2]
```diff
@@ -743,6 +743,13 @@ protected function embed_links( $data, $embed = true ) {
 						continue;
 					}
 
+					if ( empty( $request['per_page'] ) ) {
+						$matched = $this->match_request_to_handler( $request );
+						if ( ! empty( $matched ) && ! is_wp_error( $matched ) && isset( $matched[1]['args']['per_page']['maximum'] ) ) {
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/class-wp-rest-server.php]:
```suggestion
						if ( ! is_wp_error( $matched ) && isset( $matched[1]['args']['per_page']['maximum'] ) ) {
```

The empty check is superfluous.

File: tests/phpunit/tests/rest-api/rest-posts-controller.php [hunk 3]
```diff
@@ -223,6 +226,21 @@ public function test_registered_get_item_params() {
 		$this->assertSame( array( 'context', 'id', 'password' ), $keys );
 	}
 
+	public function test_registered_get_items_embed() {
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/rest-api/rest-posts-controller.php:229]: Missing `@ticket` annotation.

```suggestion
	/**
	 * @ticket 43439
	 */
	public function test_registered_get_items_embed() {
```

**Comment by spacedmonkey** [comment]
Committed.

File: src/wp-includes/rest-api/class-wp-rest-server.php [hunk 4]
```diff
@@ -743,6 +743,13 @@ protected function embed_links( $data, $embed = true ) {
 						continue;
 					}
 
+					if ( empty( $request['per_page'] ) ) {
+						$matched = $this->match_request_to_handler( $request );
```
Comment (Mamaduka) [inline src/wp-includes/rest-api/class-wp-rest-server.php:747]: @spacedmonkey, this mutates the `$request` and sets the default `context=view`, which prevents the code below from setting it to an `embed`.

This is the cause of navigation fallback regression - https://core.trac.wordpress.org/ticket/43439#comment:17.

**Pull request closed without merging**